content/post/2018-07-01-Automate-your-grammar-checks.md

```diff
@@ -0,0 +1,92 @@
+
+## Aspell? Hunspell? *spell?
+
+First of all it would be nice to find a CLI tool which could spell check plain-text files. A minute with google reveals that most frequently used spell checker tools are **Aspell** and **Hunspell**. They are similar in use and there are lots of articles on how to use them ([1](https://opensource.com/article/18/2/how-check-spelling-linux-command-line-aspell), [2](https://alexwlchan.net/2016/09/please-use-aspell/)). One problem is that they are primarily used in interactive mode and I would like to do spell checking in non-interactive mode, which produces [output unfriendly to human](https://github.com/hunspell/hunspell/blob/master/docs/hunspell.1.md#pipe-mode). Nonetheless I decided to use Hunspell in a bash wrapper for some more readable output.
+
+### Friendly Hunspell?
+
+To run Hunspell in non-interactive mode just pass text through a pipe to Hunspell program executed with `-a` flag, so simple `cat some_file | hunspell -a` will do the trick. However there is one downside to this, by default Hunspell will report its status on every line of passed text with a variety of single characters like `*`, `&`, `@`, `+`, `-`, `#`, or [others](https://github.com/hunspell/hunspell/blob/master/docs/hunspell.1.md#pipe-mode) and I'm interested only in lines starting with `&` or `#`. I also like to have some colors commands output, so let's parse Hunspell and add more colors!
+OK, to do this I came up with script below. It iterates over every file with `.md` extension and runs Hunspell on that file. Next it sorts output, removes duplicate lines and ones starting with `*` or `@(#)`. After that it parses Hunspell output to create human readable output with some colors. Unfortunately I didn't found how to get Hunspell to return different exit code than 0, so the script looks for phrase "Misspelled word" in output and returns 1 if such phrase was detected.
+
+```bash
+#!/bin/bash
+
+RET=0
+
+exec 5>&1
+for file in $(find . -name "*.md"); do
+    echo -e "\n\e[33mChecking $file\e[0m"
+    OUT=$(cat $file | \
+        hunspell -a -d en_US 2>/dev/null | \
+        sort --ignore-case | uniq | \
+        grep -v '^\*$\|@(#).*' | \
+        sed -r "s/^[&#]\s([[:alpha:]]+)\s([0-9]+)\s([0-9]+):/$(printf '\033[0m')Misspelled word $(printf '\033[31m\033[1m')\1$(printf '\033[0m') in line \2 position \3, possible alternatives:$(printf '\033[36m')/g" | \
+        tee >(cat - >&5))
+    echo -en "\e[0m"
+    if [ "$(echo $OUT | grep 'Misspelled word' | wc -l)" -gt 0 ]; then
+        RET=1
+    fi
+done
+
+exit $RET
+```
+
+### Not so markdown friendly Hunspell
+
+After creating above script and running it on some markdown files with embedded code (like this blog post) it was clear that I need some other tool. Script produced many more lines which were part of code section and weren't important, for example:
```

Comment:
Przeca wiadomo, że stworzyłeś ten skrypt. Myślę, że nie ma się co powtarzać.

After running my script on some markdown files with embedded code (like this blog post) it was clear to me that I need other tool.